From 63f0b9879b66cd5efcaa3c0a7f84f209def53b12 Mon Sep 17 00:00:00 2001
From: Rahul Ganesh <rahulgab@amazon.com>
Date: Thu, 22 Jan 2026 16:46:18 -0800
Subject: [PATCH 1/7] Remove secondstar service

EKS-A does not use the secondstar (SSH over serial) service. Disable it.

Note: tootles (metadata service, formerly hegel) is kept as EKS-A uses
it for machine metadata.

Signed-off-by: Rahul Ganesh <rahulgab@amazon.com>
---
 cmd/tinkerbell/cmd.go             | 39 +------------------
 cmd/tinkerbell/defaults.go        |  1 -
 cmd/tinkerbell/flag/global.go     |  7 ----
 cmd/tinkerbell/flag/secondstar.go | 64 -------------------------------
 cmd/tinkerbell/helpers.go         | 11 +-----
 5 files changed, 3 insertions(+), 119 deletions(-)
 delete mode 100644 cmd/tinkerbell/flag/secondstar.go

diff --git a/cmd/tinkerbell/cmd.go b/cmd/tinkerbell/cmd.go
index 4113428f..93f22ca3 100644
--- a/cmd/tinkerbell/cmd.go
+++ b/cmd/tinkerbell/cmd.go
@@ -18,7 +18,6 @@ import (
 	"github.com/tinkerbell/tinkerbell/pkg/build"
 	"github.com/tinkerbell/tinkerbell/pkg/otel"
 	"github.com/tinkerbell/tinkerbell/rufio"
-	"github.com/tinkerbell/tinkerbell/secondstar"
 	"github.com/tinkerbell/tinkerbell/smee"
 	"github.com/tinkerbell/tinkerbell/tink/controller"
 	"github.com/tinkerbell/tinkerbell/tink/server"
@@ -45,7 +44,6 @@ func Execute(ctx context.Context, cancel context.CancelFunc, args []string) erro
 		EnableTinkServer:     true,
 		EnableTinkController: true,
 		EnableRufio:          true,
-		EnableSecondStar:     true,
 		EnableUI:             true,
 		EnableCRDMigrations:  true,
 		HTTPPort:             defaultHTTPPort,
@@ -94,14 +92,6 @@ func Execute(ctx context.Context, cancel context.CancelFunc, args []string) erro
 		Config: rufio.NewConfig(rufioOpts...),
 	}
 
-	ssc := &flag.SecondStarConfig{
-		Config: &secondstar.Config{
-			SSHPort:      defaultSecondStarPort,
-			IPMITOOLPath: "/usr/sbin/ipmitool",
-			IdleTimeout:  15 * time.Minute,
-		},
-	}
-
 	uiOpts := []ui.Option{
 		ui.WithURLPrefix("/"),
 	}
@@ -119,15 +109,13 @@ func Execute(ctx context.Context, cancel context.CancelFunc, args []string) erro
 	tfs := ff.NewFlagSet("tink server - Workflow service").SetParent(hfs)
 	cfs := ff.NewFlagSet("tink controller - Workflow controller").SetParent(tfs)
 	rfs := ff.NewFlagSet("rufio - BMC controller").SetParent(cfs)
-	ssfs := ff.NewFlagSet("secondstar - SSH over serial service").SetParent(rfs)
-	uifs := ff.NewFlagSet("ui - UI service").SetParent(ssfs)
+	uifs := ff.NewFlagSet("ui - UI service").SetParent(rfs)
 	gfs := ff.NewFlagSet("globals").SetParent(uifs)
 	flag.RegisterSmeeFlags(&flag.Set{FlagSet: sfs}, s)
 	flag.RegisterTootlesFlags(&flag.Set{FlagSet: hfs}, h)
 	flag.RegisterTinkServerFlags(&flag.Set{FlagSet: tfs}, ts)
 	flag.RegisterTinkControllerFlags(&flag.Set{FlagSet: cfs}, tc)
 	flag.RegisterRufioFlags(&flag.Set{FlagSet: rfs}, rc)
-	flag.RegisterSecondStarFlags(&flag.Set{FlagSet: ssfs}, ssc)
 	flag.RegisterUIFlags(&flag.Set{FlagSet: uifs}, uic)
 	flag.RegisterGlobal(&flag.Set{FlagSet: gfs}, globals)
 	if embeddedApiserverExecute != nil && embeddedFlagSet != nil {
@@ -160,7 +148,6 @@ func Execute(ctx context.Context, cancel context.CancelFunc, args []string) erro
 		"tinkServerEnabled", globals.EnableTinkServer,
 		"tinkControllerEnabled", globals.EnableTinkController,
 		"rufioEnabled", globals.EnableRufio,
-		"secondStarEnabled", globals.EnableSecondStar,
 		"uiEnabled", globals.EnableUI,
 		"publicIP", globals.PublicIP,
 		"embeddedKubeAPIServer", globals.EmbeddedGlobalConfig.EnableKubeAPIServer,
@@ -213,14 +200,6 @@ func Execute(ctx context.Context, cancel context.CancelFunc, args []string) erro
 	// Rufio Controller
 	rc.Config.LeaderElectionNamespace = leaderElectionNamespace(inCluster(), rc.Config.EnableLeaderElection, rc.Config.LeaderElectionNamespace)
 
-	// Second star
-	if err := ssc.Convert(); err != nil {
-		return fmt.Errorf("failed to convert secondstar config: %w", err)
-	}
-	if globals.BindAddr.IsValid() {
-		ssc.Config.BindAddr = globals.BindAddr
-	}
-
 	// Initialize OTel before starting goroutines so the provider outlives
 	// all goroutines (Smee non-HTTP, consolidated HTTP server, etc.).
 	// otel.Init is a no-op when globals.OTELEndpoint is empty.
@@ -302,7 +281,7 @@ func Execute(ctx context.Context, cancel context.CancelFunc, args []string) erro
 			cliLog.Info("CRD migrations completed")
 		}
 
-		b, err := newKubeBackend(ctx, globals.BackendKubeConfig, "", globals.BackendKubeNamespace, enabledIndexes(globals.EnableSmee, globals.EnableTootles, globals.EnableTinkServer, globals.EnableSecondStar), WithQPS(globals.BackendKubeOptions.QPS), WithBurst(globals.BackendKubeOptions.Burst))
+		b, err := newKubeBackend(ctx, globals.BackendKubeConfig, "", globals.BackendKubeNamespace, enabledIndexes(globals.EnableSmee, globals.EnableTootles, globals.EnableTinkServer), WithQPS(globals.BackendKubeOptions.QPS), WithBurst(globals.BackendKubeOptions.Burst))
 		if err != nil {
 			return fmt.Errorf("failed to create kube backend: %w", err)
 		}
@@ -312,7 +291,6 @@ func Execute(ctx context.Context, cancel context.CancelFunc, args []string) erro
 		tc.Config.Client = b.ClientConfig
 		tc.Config.DynamicClient = b
 		rc.Config.Client = b.ClientConfig
-		ssc.Config.Backend = b
 		if uic.Config.EnableAutoLogin {
 			uic.Config.AutoLoginRestConfig = b.ClientConfig
 			uic.Config.AutoLoginNamespace = globals.BackendKubeNamespace
@@ -410,19 +388,6 @@ func Execute(ctx context.Context, cancel context.CancelFunc, args []string) erro
 		return nil
 	})
 
-	// SecondStar
-	g.Go(func() error {
-		if !globals.EnableSecondStar {
-			cliLog.Info("secondstar service is disabled")
-			return nil
-		}
-		ll := ternary((ssc.LogLevel != 0), ssc.LogLevel, globals.LogLevel)
-		if err := ssc.Config.Start(ctx, getLogger(ll).WithName("secondstar")); err != nil {
-			return fmt.Errorf("failed to start secondstar service: %w", err)
-		}
-		return nil
-	})
-
 	if err := g.Wait(); err != nil && !errors.Is(err, context.Canceled) {
 		return err
 	}
diff --git a/cmd/tinkerbell/defaults.go b/cmd/tinkerbell/defaults.go
index a44977fc..9fc6c861 100644
--- a/cmd/tinkerbell/defaults.go
+++ b/cmd/tinkerbell/defaults.go
@@ -15,7 +15,6 @@ import (
 const (
 	smeePublicIPInterface          = "TINKERBELL_PUBLIC_IP_INTERFACE"
 	defaultLeaderElectionNamespace = "default"
-	defaultSecondStarPort          = 2222
 	defaultHTTPPort                = 7080
 	defaultHTTPSPort               = 7443
 	defaultTinkServerPort          = 42113
diff --git a/cmd/tinkerbell/flag/global.go b/cmd/tinkerbell/flag/global.go
index aba684be..d7da6c75 100644
--- a/cmd/tinkerbell/flag/global.go
+++ b/cmd/tinkerbell/flag/global.go
@@ -25,7 +25,6 @@ type GlobalConfig struct {
 	EnableTinkServer     bool
 	EnableTinkController bool
 	EnableRufio          bool
-	EnableSecondStar     bool
 	EnableUI             bool
 	EnableCRDMigrations  bool
 	EmbeddedGlobalConfig EmbeddedGlobalConfig
@@ -64,7 +63,6 @@ func RegisterGlobal(fs *Set, gc *GlobalConfig) {
 	fs.Register(EnableTinkServer, ffval.NewValueDefault(&gc.EnableTinkServer, gc.EnableTinkServer))
 	fs.Register(EnableTinkController, ffval.NewValueDefault(&gc.EnableTinkController, gc.EnableTinkController))
 	fs.Register(EnableRufioController, ffval.NewValueDefault(&gc.EnableRufio, gc.EnableRufio))
-	fs.Register(EnableSecondStar, ffval.NewValueDefault(&gc.EnableSecondStar, gc.EnableSecondStar))
 	fs.Register(EnableUI, ffval.NewValueDefault(&gc.EnableUI, gc.EnableUI))
 	fs.Register(EnableCRDMigrations, ffval.NewValueDefault(&gc.EnableCRDMigrations, gc.EnableCRDMigrations))
 	fs.Register(LogLevelConfig, ffval.NewValueDefault(&gc.LogLevel, gc.LogLevel))
@@ -168,11 +166,6 @@ var EnableRufioController = Config{
 	Usage: "enable Rufio Controller service",
 }
 
-var EnableSecondStar = Config{
-	Name:  "enable-secondstar",
-	Usage: "enable SecondStar service",
-}
-
 var EnableUI = Config{
 	Name:  "enable-ui",
 	Usage: "enable UI service",
diff --git a/cmd/tinkerbell/flag/secondstar.go b/cmd/tinkerbell/flag/secondstar.go
deleted file mode 100644
index 68ad43d3..00000000
--- a/cmd/tinkerbell/flag/secondstar.go
+++ /dev/null
@@ -1,64 +0,0 @@
-package flag
-
-import (
-	"github.com/peterbourgon/ff/v4/ffval"
-	"github.com/tinkerbell/tinkerbell/pkg/backend/kube"
-	"github.com/tinkerbell/tinkerbell/secondstar"
-)
-
-type SecondStarConfig struct {
-	Config      *secondstar.Config
-	HostKeyPath string
-	LogLevel    int
-}
-
-var KubeIndexesSecondStar = map[kube.IndexType]kube.Index{
-	kube.IndexTypeHardwareName: kube.Indexes[kube.IndexTypeHardwareName],
-	kube.IndexTypeMachineName:  kube.Indexes[kube.IndexTypeMachineName],
-}
-
-func RegisterSecondStarFlags(fs *Set, ssc *SecondStarConfig) {
-	fs.Register(SecondStarPort, ffval.NewValueDefault(&ssc.Config.SSHPort, ssc.Config.SSHPort))
-	fs.Register(SecondStarHostKey, ffval.NewValueDefault(&ssc.HostKeyPath, ssc.HostKeyPath))
-	fs.Register(SecondStarIPMIToolPath, ffval.NewValueDefault(&ssc.Config.IPMITOOLPath, ssc.Config.IPMITOOLPath))
-	fs.Register(SecondStarIdleTimeout, ffval.NewValueDefault(&ssc.Config.IdleTimeout, ssc.Config.IdleTimeout))
-	fs.Register(SecondStarLogLevel, ffval.NewValueDefault(&ssc.LogLevel, ssc.LogLevel))
-}
-
-var SecondStarPort = Config{
-	Name:  "secondstar-port",
-	Usage: "Port to listen on for SecondStar",
-}
-
-var SecondStarHostKey = Config{
-	Name:  "secondstar-host-key",
-	Usage: "Path to the host key file for SecondStar",
-}
-
-var SecondStarIPMIToolPath = Config{
-	Name:  "secondstar-ipmitool-path",
-	Usage: "Path to the ipmitool binary",
-}
-
-var SecondStarIdleTimeout = Config{
-	Name:  "secondstar-idle-timeout",
-	Usage: "Idle timeout for SecondStar",
-}
-
-var SecondStarLogLevel = Config{
-	Name:  "secondstar-log-level",
-	Usage: "the higher the number the more verbose, level 0 inherits the global log level, a negative number disables logging",
-}
-
-func (ssc *SecondStarConfig) Convert() error {
-	// convert the host key path to an SSH Signer
-	if ssc.HostKeyPath == "" {
-		return nil
-	}
-	s, err := secondstar.HostKeyFrom(ssc.HostKeyPath)
-	if err != nil {
-		return err
-	}
-	ssc.Config.HostKey = s
-	return nil
-}
diff --git a/cmd/tinkerbell/helpers.go b/cmd/tinkerbell/helpers.go
index fe329c3d..7c775453 100644
--- a/cmd/tinkerbell/helpers.go
+++ b/cmd/tinkerbell/helpers.go
@@ -33,16 +33,13 @@ func numEnabled(globals *flag.GlobalConfig) int {
 	if globals.EnableRufio {
 		n++
 	}
-	if globals.EnableSecondStar {
-		n++
-	}
 	if globals.EnableUI {
 		n++
 	}
 	return n
 }
 
-func enabledIndexes(smeeEnabled, tootlesEnabled, tinkServerEnabled, secondStarEnabled bool) map[kube.IndexType]kube.Index {
+func enabledIndexes(smeeEnabled, tootlesEnabled, tinkServerEnabled bool) map[kube.IndexType]kube.Index {
 	idxs := make(map[kube.IndexType]kube.Index, 0)
 
 	if smeeEnabled {
@@ -58,12 +55,6 @@ func enabledIndexes(smeeEnabled, tootlesEnabled, tinkServerEnabled, secondStarEn
 			idxs[k] = v
 		}
 	}
-	if secondStarEnabled {
-		for k, v := range flag.KubeIndexesSecondStar {
-			idxs[k] = v
-		}
-	}
-
 	return idxs
 }
 
-- 
2.46.0

